xds/internal/client/v2client_rds.go

```diff
@@ -94,42 +95,154 @@ func generateRDSUpdateFromRouteConfiguration(rc *xdspb.RouteConfiguration, host
+
+	routes, err := routesProtoToSlice(vh.Routes)
+	if err != nil {
+		return rdsUpdate{}, fmt.Errorf("received route is invalid: %v", err)
 	}
+	return rdsUpdate{routes: routes}, nil
+}
 
-	if wc := route.GetWeightedClusters(); wc != nil {
-		m, err := weightedClustersProtoToMap(wc)
-		if err != nil {
-			return rdsUpdate{}, fmt.Errorf("matched weighted cluster is invalid: %v", err)
+func routesProtoToSlice(routes []*routepb.Route) ([]*Route, error) {
+	var routesRet []*Route
+
+	for _, route := range routes {
+		match := route.GetMatch()
+		if match == nil {
+			return nil, fmt.Errorf("route %v doesn't have a match", route)
 		}
-		return rdsUpdate{weightedCluster: m}, nil
-	}
 
-	// When there's just one cluster, we set weightedCluster to map with one
-	// entry. This mean we will build a weighted_target balancer even if there's
-	// just one cluster.
-	//
-	// Otherwise, we will need to switch the top policy between weighted_target
-	// and CDS. In case when the action changes between one cluster and multiple
-	// clusters, changing top level policy means recreating TCP connection every
-	// time.
-	return rdsUpdate{weightedCluster: map[string]uint32{route.GetCluster(): 1}}, nil
+		if len(match.GetQueryParameters()) != 0 {
+			// Ignore route with query parameters.
+			continue
+		}
+
+		if caseSensitive := match.GetCaseSensitive(); caseSensitive != nil && !caseSensitive.Value {
+			return nil, fmt.Errorf("route %v has case-sensitive false", route)
+		}
+
+		var routeTemp Route
+
+		pathSp := match.GetPathSpecifier()
+		if pathSp == nil {
+			return nil, fmt.Errorf("route %v doesn't have a path specifier", route)
+		}
+		switch pt := pathSp.(type) {
+		case *routepb.RouteMatch_Prefix:
+			routeTemp.Prefix = &pt.Prefix
+		case *routepb.RouteMatch_Path:
+			routeTemp.Path = &pt.Path
+		case *routepb.RouteMatch_SafeRegex:
+			routeTemp.Regex = &pt.SafeRegex.Regex
+		case *routepb.RouteMatch_Regex:
+			return nil, fmt.Errorf("route %v has Regex, expected SafeRegex instead", route)
+		}
+
+		for _, header := range match.GetHeaders() {
+			var headerTemp HeaderMatcher
+			switch ht := header.GetHeaderMatchSpecifier().(type) {
+			case *routepb.HeaderMatcher_ExactMatch:
+				headerTemp.ExactMatch = &ht.ExactMatch
+			case *routepb.HeaderMatcher_SafeRegexMatch:
+				headerTemp.RegexMatch = &ht.SafeRegexMatch.Regex
+			case *routepb.HeaderMatcher_RangeMatch:
+				headerTemp.RangeMatch = &Int64Range{
+					Start: ht.RangeMatch.Start,
+					End:   ht.RangeMatch.End,
+				}
+			case *routepb.HeaderMatcher_PresentMatch:
+				headerTemp.PresentMatch = &ht.PresentMatch
+			case *routepb.HeaderMatcher_PrefixMatch:
+				headerTemp.PrefixMatch = &ht.PrefixMatch
+			case *routepb.HeaderMatcher_SuffixMatch:
+				headerTemp.SuffixMatch = &ht.SuffixMatch
+			case *routepb.HeaderMatcher_RegexMatch:
+				return nil, fmt.Errorf("route %v has a header matcher with Regex, expected SafeRegex instead", route)
+			default:
+				continue
```

Comment:
Does the `default` only cover the case where `ht == nil`? Just wondering if we need a log statement here.

Reply:
Or new types that may be added later. Added a log.